### PR TITLE
Additions and fixes for Fields and FieldFunctions

### PR DESCRIPTION
**Pull request by akva2**
This adds support for
- mixed field functions (as in one field with components using different bases - like for Raviart-Thomas in Stokes)
- ability to evaluate hessians for fields
- additional field function types - gradients and laplacians for scalar and vectorial fields

The motivation for this is to be able to use a high fidelity finite element solution as an exact solution in a coarser simulation. In particular it allows for estimating the efficiency of aposteriori error estimators.

Quite a big PR but it's hard to split it more than this. As part of this is significantly reduced code duplication in the field classes.

### CODE REVIEW

**Comment by akva2** [comment]
jenkins build this with downstreams please

**Comment by akva2** [comment]
jenkins build this with downstreams simra-postproc=9 please

**Comment by akva2** [comment]
jenkins build this with downstreams simra-postproc=9 please

**Comment by akva2** [comment]
jenkins build this with downstreams simra-postproc=9 please

**Review by kmokstad — COMMENTED** [review]
You're right. This PR is quite large, and not possible to grasp all details here. But at least some minor things were spotted (mostly cosmetic issues).

File: src/ASM/ASMs2Dmx.C [hunk 1]
```diff
@@ -72,6 +73,28 @@ Go::SplineCurve* ASMs2Dmx::getBoundary (int dir, int basis)
...
+
+  if (m_basis.empty()) {
+    m_basis.resize(nfx.size());
+    nb.resize(nfx.size());
+  }
+
+  Go::ObjectHeader head;
+  m_basis[basis-1] = std::make_shared<Go::SplineSurface>();
+  is >> head >> *m_basis[basis-1];
+  nb[basis-1]= m_basis[basis-1]->numCoefs_u()*m_basis[basis-1]->numCoefs_v();
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/ASMs2Dmx.C:92]: done

File: src/ASM/ASMs2Dmx.C [hunk 2]
```diff
@@ -72,6 +73,28 @@ Go::SplineCurve* ASMs2Dmx::getBoundary (int dir, int basis)
...
+{
+  if (basis == 0)
+    return this->ASMs2D::read(is);
+
+  if (basis < 0 || basis > static_cast<int>(nfx.size()))
+    return false;
+
+  if (m_basis.empty()) {
+    m_basis.resize(nfx.size());
+    nb.resize(nfx.size());
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/ASMs2Dmx.C:86]: vectors default resize to 0. m_basis is a vector of shared pointers which default initialize to nullptr.

File: src/ASM/ASMsupel.h [hunk 3]
```diff
@@ -39,7 +39,7 @@ class ASMsupel : public ASMbase
   virtual ~ASMsupel() {}
 
   //! \brief Creates an instance by reading the given input stream.
-  virtual bool read(std::istream& is);
+  virtual bool read(std::istream& is, int = 0);
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/ASMsupel.h:42]: fixed.

File: src/ASM/LR/LRSplineFields3Dmx.C [hunk 4]
```diff
@@ -48,6 +49,15 @@ bool LRSplineFields3Dmx::valueNode (size_t node, Vector& vals) const
 }
 
 
+bool LRSplineFields3Dmx::valueCoor (const Vec4& x, Vector& vals) const
+{
+  if (x.u)
+    return this->valueFE(ItgPoint(x.u[0],x.u[1],x.u[2]),vals);
+
+  return false;
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/LR/LRSplineFields3Dmx.C:59]: okay

File: src/Utility/FieldFunctions.C [hunk 5]
```diff
@@ -120,39 +124,63 @@ bool FieldFuncHDF5::load (const std::vector<std::string>& fieldNames,
...
+        else
+          patch[ip] = nullptr;
+        if (patch[ip])
+        {
+          std::stringstream strg2(g2);
+          patch[ip]->read(strg2);
+        }
+        else
+          std::cerr <<" *** FieldFuncHDF5::load: Undefined basis "<< sbasis.str()
+                   <<" ("<< g2.substr(0,9) <<")"<< std::endl;
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/Utility/FieldFunctions.C:190]: problem is it's same same, but different, ie as the ASMxD::create calls have different parameters. I tried to do it with passing a lambda to parameter of a second lambda but it was not prettier.

File: src/Utility/Test/TestFieldFunctions.C [hunk 6]
```diff
@@ -89,6 +95,9 @@ TEST(TestFieldFunctions, 3D1P)
   VecFieldFunction f3D_vec("src/Utility/Test/refdata/Field3D-1P",
                            "Stokes-1", "u", 0);
 
+  VecFieldFunction f3Dmx_vec("src/Utility/Test/refdata/Field3D-1Pmx",
+                           "Stokes", "u_x|u_y|u_z", 0);
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/Utility/Test/TestFieldFunctions.C]: fixed in a later commit.

File: src/ASM/LR/LRSplineField.h [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+//!
+//! \brief Utility class for LR spline-based finite element fields.
+//!
+//==============================================================================
+
+#ifndef _LRSPLINE_FIELD_H
+#define _LRSPLINE_FIELD_H
+
+#include "MatVec.h"
+#include "matrixnd.h"
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/LR/LRSplineField.h:18]: okay

File: src/ASM/SplineFields3D.C [hunk 8]
```diff
@@ -64,7 +66,7 @@ SplineFields3D::SplineFields3D (const Go::SplineVolume* svol,
                                 const RealArray& v, int cmp, const char* name)
   : Fields(name), basis(svol), vol(svol)
 {
-  nsd = 3;
+  nsd = svol->dimension();
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/SplineFields3D.C:69]: Not new code, but I can remove it if you prefer.

File: src/Utility/FieldFunctions.h [hunk 9]
```diff
@@ -30,6 +30,9 @@ class ProcessAdm;
 
 class FieldFuncBase
 {
+public:
+  static constexpr int FIXED_LEVEL = 1 << 24; //!< Bit flag in level for using fixed
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/Utility/FieldFunctions.h:34]: see skype conversation. constexpr is for compile time constants.

File: src/Utility/FieldFunctions.h [hunk 10]
```diff
@@ -87,11 +90,13 @@ class FieldFuncHDF5 : public FieldFuncBase
   //! \brief Clears the field container.
   virtual void clearField() = 0;
 
+  bool hasMultipleLevels; //!< True if we have multiple time levels
+  mutable int    lastLevel; //!< The last time level read from
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/Utility/FieldFunctions.h:94]: done.

File: src/ASM/LR/ASMu2Dmx.h [hunk 11]
```diff
@@ -47,6 +47,8 @@ class ASMu2Dmx : public ASMu2D, private ASMmxBase
   // Methods for model generation
   // ============================
 
+  //! \brief Creates an instance by reading the given input stream.
+  virtual bool read(std::istream&, int basis = 0);
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/LR/ASMu2Dmx.h:51]: Removed where not necessary, but there are calls to ASMs1D::read, ASMs2D::read, ASMs3D::read, ASMu2D::read and ASMu3D::read in their respective contexts.

File: src/ASM/LR/LRSplineField.C [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+    surf.computeBasis(x.u,x.v,spline,iel);
+    SplineUtils::extractBasis(spline, N, dNdu);
+  }
+
+  dNdX.multiply(dNdu,Jac); // dNdX = dNdu * Jac
+
+  if (Hess && !utl::Hessian(*Hess,*d2NdX2,Jac,Xnod,d2Ndu2,dNdX,false))
+    return false;
+
+  return true;
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/LR/LRSplineField.C:95]: not sure what it is the same as.

File: src/ASM/LR/LRSplineField.C [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+  }
+
+  // Evaluate the Jacobian inverse
+  if (!utl::Jacobian(Jac,dNdX,Xnod,dNdu))
+    return false;
+
+  if (Hess && !utl::Hessian(*Hess,*d2NdX2,Jac,Xnod,d2Ndu2,dNdX,true))
+    return false;
+
+  return true;
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/LR/LRSplineField.C:138]: same as what?

File: src/ASM/LR/LRSplineField.C [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+    vol.computeBasis(x.u,x.v,x.w,spline,iel);
+    SplineUtils::extractBasis(spline, N, dNdu);
+  }
+
+  dNdX.multiply(dNdu,Jac); // dNdX = dNdu * Jac
+
+  if (Hess && !utl::Hessian(*Hess,*d2NdX2,Jac,Xnod,d2Ndu2,dNdX,false))
+    return false;
+
+  return true;
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/LR/LRSplineField.C:171]: same as what?

File: src/ASM/LR/LRSplineField3D.C [hunk 15]
```diff
@@ -65,18 +67,19 @@ double LRSplineField3D::valueFE (const ItgPoint& x) const
...
 
 
 double LRSplineField3D::valueCoor (const Vec4& x) const
 {
-  // Just produce a segmentation fault, if invoked without the parameters
-  return this->valueFE(ItgPoint(x.u[0],x.u[1],x.u[2]));
+  if (x.u)
+    return this->valueFE(ItgPoint(x.u[0],x.u[1],x.u[2]));
+
+  return false;
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/LR/LRSplineField3D.C:82]: done.

File: src/ASM/LR/LRSplineField.C (hunk 12 above)
[the remark this answers is not in the document]
Comment (kmokstad) [reply src/ASM/LR/LRSplineField.C:95]: ok, original comment appears to been lost. But the idea was that L92-95 could be shortened into
```return Hess ? utl::Hessian(*Hess,*d2NdX2,Jac,Xnod,d2Ndu2,dNdX,false) : true;```

But no big deal.

[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/LR/LRSplineField.C:95]: done.

File: src/ASM/LR/LRSplineField.C (hunk 14 above)
[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/LR/LRSplineField.C:171]: done.

File: src/ASM/LR/LRSplineField.C (hunk 13 above)
[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/LR/LRSplineField.C:138]: done

**Comment by akva2** [comment]
jenkins build this with downstreams simra-postproc=9 please

**Comment by akva2** [comment]
jenkins build this with downstreams simra-postproc=9 please

File: src/ASM/ASMs2Dmx.C (hunk 2 above)
[the remark this answers is not in the document]
Comment (kmokstad) [reply src/ASM/ASMs2Dmx.C:86]: Ok, I thought the default constructor of the template class was being used if no second argument. And the default constructor of an `int` isn't that unininialized variable, at least when compiled optimized?

[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/ASMs2Dmx.C:86]: yeah, you are of course correct, i was confused. remedied.

**Comment by akva2** [comment]
jenkins build this with downstreams simra-postproc=9 please

**Comment by akva2** [comment]
jenkins build this with downstreams simra-postproc=9 please

**Comment by akva2** [comment]
jenkins build this with downstreams simra-postproc=9 please

**Pull request merged**